[Marciano Renzo William] iP

midnightfeverrr opened the pull request:
# Pull Request
My very own pull request, pls give an A 💯

## Quote
> What if one day google got deleted, and we could not google what happened to google?

## What's Good 👍 
* Find task by keyword feature
* Save tasks to a file, so you can quit the program, and **boom** your tasks are stil there (*click* **nice**)

## Things to Note 📓 
1. This program is a little bit buggy
2. Test cases are not accurate 🗡️ 

## Stunning Stuffs 😧 
`Parser.java` line 25-31 where I implemented the usage of enum:
```ruby
/**
 * Class to represent a parser.
 */
public class Parser {
    // usage of Enum
    private enum Commands {
        BYE, LIST, MARK, UNMARK, TODO, DEADLINE, EVENT, DELETE, FIND
    }
    private static final DateTimeFormatter TIME_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
    private static Ui ui = new Ui();
```
This IP is fun ! ~~Java isn't hehe~~

## Todo ✅ 
- [x] Write markdowns
- [ ] Continue writing IP Code

## Video Link
[Here's](https://www.youtube.com/watch?v=dQw4w9WgXcQ) a video of my project.

Commit:
Add Gradle support

midnightfeverrr committed:
Level-1

midnightfeverrr committed:
Duke Level-2: Add, List

midnightfeverrr committed:
Duke Level-3: Mark as Done

midnightfeverrr committed:
Duke Level-4: ToDo, Event, Deadline

midnightfeverrr committed:
Duke A-TextUiTesting: Automated Text UI Testing

midnightfeverrr committed:
Duke Level-5: Handle Errors

midnightfeverrr committed:
Duke Level-6: Delete

midnightfeverrr committed:
Duke A-Enums: Use Enums

midnightfeverrr committed:
add storage to save task list

midnightfeverrr committed:
use appropriate format of date and time

midnightfeverrr committed:
Merge branch 'branch-Level-7'

midnightfeverrr committed:
Merge branch 'branch-Level-8'

midnightfeverrr committed:
Update storage class to match date and time format

midnightfeverrr committed:
fix date and time formatting

divide classes to packages

midnightfeverrr committed:
Merge branch 'add-gradle-support'

midnightfeverrr committed:
add gradle support

midnightfeverrr committed:
add junit test cases

midnightfeverrr committed:
add jar file

midnightfeverrr committed:
add and fix JavaDoc

midnightfeverrr committed:
follow the coding standard

midnightfeverrr committed:
add find command

midnightfeverrr committed:
Merge branch 'branch-A-JavaDoc'

midnightfeverrr committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/Duke/Processor/Parser.java
#	src/main/java/Duke/UI.java
#	src/main/java/duke/Duke.java
#	src/main/java/duke/processor/TaskList.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Delete.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/List.java
#	src/main/java/duke/task/Mark.java
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/Todo.java

midnightfeverrr committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/processor/Parser.java
#	src/main/java/duke/processor/Storage.java
#	src/main/java/duke/processor/TaskList.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Delete.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/List.java
#	src/main/java/duke/task/Mark.java
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/Todo.java
#	src/main/java/duke/task/Unmark.java

midnightfeverrr committed:
fix file names

midnightfeverrr committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/processor/Parser.java
#	src/main/java/duke/processor/Storage.java
#	src/main/java/duke/processor/TaskList.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Delete.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/List.java
#	src/main/java/duke/task/Mark.java
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/Todo.java
#	src/main/java/duke/task/Unmark.java

midnightfeverrr committed:
Change bot's name

clarence-chew commented on src/main/java/ava/Ava.java:
```diff
@@ -0,0 +1,61 @@
+ */
+public class Ava {
+    private Ui ui;
+    private TaskList tasklist;
```
taskList could be an alternative name you can consider for this variable, consistently viewing it as two words.

clarence-chew commented on src/main/java/ava/Ui.java:
```diff
@@ -0,0 +1,118 @@
+     * @param tasks TaskList.
+     * @param numOfTask Number of Tasks.
+     */
+    public void showAddOnTask(TaskList tasks, int numOfTask) {
```
numOfTasks seems to be data that one can get from the TaskList - perhaps we can discuss on advantages of separating it from the TaskList.

clarence-chew commented on src/main/java/ava/exception/NoDescriptionException.java:
```diff
@@ -0,0 +1,15 @@
+ */
+public class NoDescriptionException extends Exception {
+    /**
+     * The Constructor for NoCommandException.
```
I believe this is a small oversight here with the class name.

clarence-chew commented on src/main/java/ava/processor/Parser.java:
```diff
@@ -0,0 +1,217 @@
+        try {
+            try {
+                command = Parser.Commands.valueOf(chat.toUpperCase().split(" ")[0]);
+            } catch (Exception e) {
```
Perhaps this exception could be further specified as an IllegalArgumentException? I noticed this recurrence of catching Exception, instead of a more specific type, in other places too.

clarence-chew commented on src/main/java/ava/processor/TaskList.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    /**
+     * Adds a specific tasks to the TaskList.
+     *
+     * @param t A specific tasks to add.
```
"tasks" might be better as singular "task" here, since we are dealing with one task. I noticed this in the other functions as well. From my understanding, it could have arose from a refactor of "task" -> "tasks".

clarence-chew commented on src/main/java/ava/task/Event.java:
```diff
@@ -0,0 +1,83 @@
+     * The constructor for Event task.
+     *
+     * @param description Task description.
+     * @param isDone Task state.
```
This seems to be missing a @param for `time`.

clarence-chew commented on src/main/java/ava/task/Delete.java:
```diff
@@ -0,0 +1,34 @@
+    private int num;
+
+    /**
+     * The constructor for Delete task
```
This seems to be missing a @param for `num`.

clarence-chew commented on src/main/java/ava/task/Deadline.java:
```diff
@@ -0,0 +1,84 @@
+     * The constructor for Deadline task.
+     *
+     * @param description Task description.
+     * @param isDone Task state.
```
This seems to be missing a @param for `time`.

clarence-chew commented on src/main/java/ava/processor/TaskList.java:
```diff
@@ -0,0 +1,93 @@
+    /**
+     * The constructor for TaskList with an input of ArrayList.
+     *
+     * @param task An ArrayList of tasks.
```
This parameter might be better as plural since it is an ArrayList.

midnightfeverrr replied:
Thanks for pointing out many errors in the documentations ! :D

clarence-chew commented on src/main/java/ava/processor/Storage.java:
```diff
@@ -0,0 +1,84 @@
+    /**
+     * Writes tasks with a specified format to the file.
+     *
+     * @param task An ArrayList of tasks to write.
```
This parameter `task` might be better as `tasks` as it is an ArrayList.

clarence-chew commented on src/main/java/ava/task/Task.java:
```diff
@@ -0,0 +1,95 @@
+ * Abstract class to represent the tasks.
+ */
+public abstract class Task {
+    public Boolean isBye;
```
This class variable might be better as `protected`, if access is required - consider writing an accessor?

midnightfeverrr replied:
I changed it, but haven't commited. Thanks !

clarence-chew commented on src/main/java/ava/task/Task.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    /**
+     * Overridden toString method for task details.
```
One may consider writing this as `Obtains a string representation of the task for printing to console.` (or other alternatives) to keep in line with the first word being a verb.

clarence-chew commented on src/main/java/ava/task/Task.java:
```diff
@@ -0,0 +1,95 @@
+     */
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+
+    };
```
This semicolon seems to be an small oversight.

clarence-chew commented on src/main/java/ava/task/Todo.java:
```diff
@@ -0,0 +1,76 @@
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        tasks.add(this);
+        ui.showAddOnTask(tasks, (tasks.size() - 1));
+        storage.write(tasks.getTasks());
```
Interesting idea to have the chatbot to save the tasks to defend against the case of unexpected program termination.

clarence-chew reviewed (commented):
Great job with your chatbot - here's some coding standard nits to think about.

midnightfeverrr committed:
Implements a GUI using javaFX technology

midnightfeverrr committed:
Merge branch 'branch-Level-10'

RichDom2185 commented on src/main/java/ava/Ui.java:
```diff
@@ -0,0 +1,64 @@
+    /**
+     * Shows the added task.
+     *
+     * @param tasks TaskList.
```
Could the documentation be clearer?

RichDom2185 commented on src/main/java/ava/Ui.java:
```diff
@@ -0,0 +1,64 @@
+     * @param numOfTask Number of Tasks.
+     * @return Message printed when task is added.
+     */
+    public String showAddOnTask(TaskList tasks, int numOfTask) {
```
I think that since this method returns a `String`, would it be better if the method was named `getAddOnTaskMessage` instead?

RichDom2185 commented on src/main/java/ava/exception/NoCommandException.java:
```diff
@@ -0,0 +1,15 @@
+package ava.exception;
+
+/**
+ * Class to represent Duke.Exception.NoCommandException.
```
Is this a typo in the documentation? Should this be `Ava.Exception.NoCommandException`?

```suggestion
 * Class to represent Ava.Exception.NoCommandException.
```

RichDom2185 commented on src/main/java/ava/exception/NoDescriptionException.java:
```diff
@@ -0,0 +1,15 @@
+package ava.exception;
+
+/**
+ * Class to represent Duke.Exception.NoDescriptionException.
```
Ditto. Is this a typo with the documentation?

```suggestion
 * Class to represent Ava.Exception.NoDescriptionException.
```

RichDom2185 commented on src/main/java/ava/exception/NoTimeException.java:
```diff
@@ -0,0 +1,15 @@
+package ava.exception;
+
+/**
+ * Class to represent Duke.Exception.NoTimeException.
```
Ditto with the documentation. There are also other places in the code where this seems to be occurring. Could you perhaps change this as well?

RichDom2185 commented on src/main/java/ava/processor/Parser.java:
```diff
@@ -0,0 +1,219 @@
+     *
+     * @param chat Input from the scanner.
+     * @param tasklist ArrayList of tasks.
+     * @return Task object
```
Is this a typo?

```suggestion
     * @return Task object.
```

RichDom2185 commented on src/main/java/ava/processor/Parser.java:
```diff
@@ -0,0 +1,219 @@
+            switch (command) {
+            case BYE:
+                return new Bye();
+
+            case LIST:
+                return new List();
+
+            case UNMARK:
+                return unmark(chat, tasklist);
+
+            case MARK:
+                return mark(chat, tasklist);
+
+            case TODO:
+                return addTask(chat, tasklist);
+
+            case DEADLINE:
+                return addTask(chat, tasklist);
+
+            case EVENT:
+                return addTask(chat, tasklist);
+
+            case DELETE:
+                return delete(chat, tasklist);
+
+            case FIND:
+                return find(chat, tasklist);
+
```
Are these empty lines necessary?
```suggestion
            case LIST:
                return new List();
            case UNMARK:
                return unmark(chat, tasklist);
            case MARK:
                return mark(chat, tasklist);
            case TODO:
                return addTask(chat, tasklist);
            case DEADLINE:
                return addTask(chat, tasklist);
            case EVENT:
                return addTask(chat, tasklist);
            case DELETE:
                return delete(chat, tasklist);
            case FIND:
                return find(chat, tasklist);
```

RichDom2185 reviewed (commented):
Overall, I really like how thorough you were with your exceptions. Good job!

midnightfeverrr committed:
Fix exceptions

midnightfeverrr committed:
delete assert implementation

midnightfeverrr committed:
fix bug

midnightfeverrr committed:
Change build gradle files to release JAR files

midnightfeverrr committed:
Use assertions before returning response from Ava

midnightfeverrr committed:
Minor bug fix and code cleanup.

Reviewers left some comments on the PR, the majority with regards to Javadocs, whitespaces, and minor naming issues.

Fix typos inside the Javadocs in every public and private method. Delete several methods with unused parameters due to copy-pasting of the similar methods. Remove whitespaces betweeen switch cases. Merge several switch cases to improve readability. Refactor several variable naming to follow coding style.

My intention in this commit is to make my code more readable. This commit should resolve some problems regarding my code style pointed out by the reviewers.

Changes made as commits:
* Add missing return Javadocs

* Fix typos in Javadocs

* Add missing punctuation marks

* Add Javadocs for exceptions

* Delete unused parameters

* Delete excess whitespaces

* Refactor variable names

* Merge switch cases

midnightfeverrr committed:
Implement toStringArray() method

This method uses Java Stream to convert tasks in the form of TaskList into an array
of its string representation by mapping each object to its respective toString() method.

The method is inspired by my senior @Huizhuansam.

midnightfeverrr committed:
Merge pull request #1 from midnightfeverrr/branch-A-Assertions

Use assertions before returning response from Ava

midnightfeverrr committed:
Merge pull request #2 from midnightfeverrr/branch-A-CodeQuality

Minor bug fix and code cleanup.

midnightfeverrr committed:
Merge pull request #3 from midnightfeverrr/branch-A-Streams

Implement toStringArray() method

midnightfeverrr committed:
add CI
fix bug regarding find function

midnightfeverrr committed:
Implemented Sort function

Refactor code names

Refine OOP

midnightfeverrr committed:
fix yml

midnightfeverrr committed:
Fix yml files

Add a new exception to handle wrong index input

midnightfeverrr committed:
fix test cases

midnightfeverrr committed:
fix checkstyle errors

midnightfeverrr committed:
Add User Guide

midnightfeverrr committed:
Minor changes

midnightfeverrr committed:
Change header comments

midnightfeverrr committed:
Change naming format
